src/event/value.rs

```diff
@@ -1,41 +1,96 @@
-use crate::{event::timestamp_to_string, Result};
+use crate::event::{string_to_timestamp, timestamp_to_string};
 use bytes::Bytes;
 use chrono::{DateTime, Utc};
 use derive_is_enum_variant::is_enum_variant;
-use serde::{Serialize, Serializer};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
 use std::collections::BTreeMap;
 use std::convert::{TryFrom, TryInto};
 use std::iter::FromIterator;
 use toml::value::Value as TomlValue;
 
-#[derive(PartialEq, Debug, Clone, is_enum_variant)]
+#[derive(PartialEq, Debug, Clone, is_enum_variant, Deserialize, Serialize)]
+#[serde(untagged)]
 pub enum Value {
-    Bytes(Bytes),
+    // Order of variants is important here, because for an untagged enum deserialization is
+    // attempted top to bottom, so when types can be confused for eachother we want the more
+    // specific variant listed first.
     Integer(i64),
     Float(f64),
     Boolean(bool),
-    Timestamp(DateTime<Utc>),
     Map(BTreeMap<String, Value>),
     Array(Vec<Value>),
+    #[serde(
+        serialize_with = "serialize_as_string_timestamp",
+        deserialize_with = "deserialize_as_string_timestamp"
+    )]
+    Timestamp(DateTime<Utc>),
+    #[serde(
+        serialize_with = "serialize_as_string_bytes",
+        deserialize_with = "deserialize_as_string_bytes"
+    )]
+    Bytes(Bytes),
+    #[serde(
```

Comment:
Similarly, I don't think this PR is the right place to be making changes to our core data model deserialization and this is going to impact existing work in this area.